[LangOptions] Remove the option to enable/disable implicit existential opening.

hborla opened the pull request:
This feature has been accepted in Swift evolution, so there's no need to have a flag to enable/disable it. This also fixes an issue where the feature was not enabled by default in LLDB.

This change also uses opaque value placeholders where code completion was previously sending a pre-type-checked AST through the constraint system, which tripped an assertion during `shouldOpenExistentialCallArgument` when looking up types for AST nodes that should have been cached in the constraint system.

Resolves: rdar://94730905

hborla committed:
[LangOptions] Remove the option to enable/disable implicit existential
opening.

hborla commented:
@swift-ci please smoke test

slavapestov commented on test/Generics/existential_restrictions.swift:
```diff
@@ -8,14 +8,14 @@ protocol CP : class { }
 }
 
-func fP<T : P>(_ t: T) { }
+func fP<T : P>(_ t: T?) { }
```
Is the addition of '?' in this file intentional?

hborla replied:
Yes, I made these parameters optional so the diagnostics stay the same later in the file (instead of disappearing due to opening)

hborla committed:
[CodeCompletion] Use an opaque value placeholder for operator arguments
in code completion to avoid sending a pre-type-checked AST through the
constraint system.

ahoppen commented on lib/Sema/TypeCheckCodeCompletion.cpp:
```diff
@@ -482,6 +482,11 @@ TypeChecker::getTypeOfCompletionOperator(DeclContext *DC, Expr *LHS,
+  // Use a placeholder expr for the LHS argument to avoid sending
+  // a pre-type-checked AST through the constraint system.
+  OpaqueValueExpr argExpr(LHS->getSourceRange(), LHSTy,
+                          /*isPlaceholder=*/true);
```
Yes, we only care about the type. We might even be able to get rid of the `ConstraintSystemFlags::ReusePrecheckedType` option above. Not entirely sure though.
